test/run.py

```diff
@@ -0,0 +1,248 @@
+         *controller_tests_parameters],
+        fail_on_failure=False)
+    return subp.returncode == 0
+
+
+def get_end_to_end_tests(
+        command_line_specified_tests: Optional[List[str]] = None) -> [str]:
+    if command_line_specified_tests is None \
+            or len(command_line_specified_tests) == 0:
+        end_to_end_tests = []
+        # Run all tests, except in directories named "*.ignored"
+        for tdir in glob(os.path.join(TEST_DIRECTORY, 'end-to-end', '*', '*')):
+            if os.path.isfile(os.path.join(tdir, 'plz.config.json')) \
+                    and '.ignored' not in tdir:
+                end_to_end_tests.append(
+                    # Remove the dir path until end-to-end
+                    tdir[len(os.path.join(TEST_DIRECTORY, '')):])
+        return end_to_end_tests
+    # Run selected tests.
+    # But first, verify all arguments are actually test directories
+    for tdir in command_line_specified_tests:
+        # Remove trailing slash if present
+        if tdir.endswith(os.sep):
+            tdir = tdir[:-len(os.sep)]
+        if re.search(r'^end-to-end/[A-Za-z0-9\-]+/[A-Za-z0-9\-]+$',
+                     tdir) is None \
+                or not os.path.isfile(
+                    os.path.join(
+                        TEST_DIRECTORY, tdir, 'plz.config.json')):
+            raise ValueError(f'{tdir} is not a test directory')
+    return command_line_specified_tests
+
+
+def _run_test_piping_output(
+        run_end_to_end_test_args: dict, output_pipe: multiprocessing.Pipe()):
+    os.dup2(output_pipe.fileno(), sys.stdout.fileno())
+    run_end_to_end_test(**run_end_to_end_test_args)
+
+
+def run_end_to_end_tests(
+        network: str, plz_host: str, plz_port: int, bless: bool,
+        in_parallel: bool,
+        command_line_specified_tests: Optional[List[str]] = None) -> Set[str]:
+    end_to_end_tests = get_end_to_end_tests(command_line_specified_tests)
+
+    if in_parallel:
+        pool = multiprocessing.Pool(processes=len(end_to_end_tests))
+        pipe_pairs = [multiprocessing.Pipe() for _ in end_to_end_tests]
+        process_args = [
+            (
+                {
+                    'network': network,
+                    'plz_host': plz_host,
+                    'plz_port': plz_port,
+                    'bless': bless,
+                    'test_name': end_to_end_tests[i]
+                },
+                output_pipe
+            )
+            for i, (_, output_pipe) in enumerate(pipe_pairs)]
+        readers = [os.fdopen(r.fileno(), 'r') for r, _ in pipe_pairs]
+        pool_result = pool.starmap_async(
+            _run_test_piping_output,
+            process_args)
+
+        result_ready = False
+        while not result_ready:
+            if pool_result.ready():
+                result_ready = True
+            for reader in readers:
+                while True:
+                    line = reader.readline()
+                    if len(line) == 0:
```

Comment:
Juse use `if not line`.

There are a lot of tests that check `if len(something) == 0`, `if len(something) != 0`, or `if len(something) > 0`. These can typically be simplified.